File: packages/react-native-airfob/ios/AirfobRemediation.swift
import Foundation
import UIKit

/// Turns a failing diagnostic into a tap that fixes it — as far as iOS allows.
///
/// **This is where iOS and Android diverge sharply.** Android can deep-link to
/// the exact settings screen for Bluetooth, location, notifications, or battery
/// optimisation. iOS offers exactly one public destination:
/// `UIApplication.openSettingsURLString`, this app's own settings page.
///
/// The `App-Prefs:root=Bluetooth` style URLs that circulate are private API. They
/// work in development and get apps rejected at review, so they are not used
/// here. Every action id therefore lands on the same screen, and the difference
/// is carried by the guidance text in the check's `remedy` field.
///
/// Practical consequence: iOS diagnostics must *say more*, because they can *do*
/// less. Keep the remedy strings specific — "Settings > Bluetooth", not
/// "check your settings".
enum AirfobRemediation {

    // Shared with Android and the JS layer so one Mendix page renders either.
    static let openBluetooth = "openBluetoothSettings"
    static let openAppSettings = "openAppSettings"
    static let openLocation = "openLocationSettings"
    static let openNotifications = "openNotificationSettings"
    static let openBattery = "openBatterySettings"
    static let requestBatteryExemption = "requestBatteryExemption"

    private static let known: Set<String> = [
        openBluetooth, openAppSettings, openLocation,
        openNotifications, openBattery, requestBatteryExemption
    ]

    /// - Returns: true when a settings screen was opened.
    /// - Throws: `AirfobError` for an unknown id, so callers surface it rather
    ///   than leaving a dead button.
    @discardableResult
    static func run(_ actionId: String) throws -> Bool {
        guard known.contains(actionId) else {
            throw AirfobError("E_SDK", "Unknown remediation action \"\(actionId)\"")
        }

        guard let url = URL(string: UIApplication.openSettingsURLString) else {
            throw AirfobError("E_SDK", "Settings URL is unavailable")
        }

        // Battery optimisation has no iOS equivalent at all — Low Power Mode is
        // system-wide and cannot be exempted per app. Say so rather than opening
        // a settings page that has nothing relevant on it.
        if actionId == openBattery || actionId == requestBatteryExemption {
            AirfobLog.shared.write(
                "info", "perm", "REMEDIATE_NOOP",
                "No per-app battery exemption exists on iOS"
            )
            throw AirfobError(
                "E_SDK",
                "iOS has no per-app battery exemption. Ask the user to turn off Low Power Mode."
            )
        }

        var opened = false
        let semaphore = DispatchSemaphore(value: 0)

        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:]) { success in
                opened = success
                semaphore.signal()
            }
        }
        _ = semaphore.wait(timeout: .now() + 3)

        AirfobLog.shared.write(
            opened ? "info" : "error",
            "perm",
            opened ? "REMEDIATE" : "REMEDIATE_FAIL",
            opened ? "Opened app settings for \(actionId)" : "Could not open settings"
        )

        return opened
    }
}

File: packages/react-native-airfob/android/src/main/java/com/airfob/AirfobRemediation.kt
package com.airfob

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings

/**
 * Turns a failing diagnostic into a tap that fixes it.
 *
 * The difference between "Bluetooth is off" and a button that opens Bluetooth
 * settings is the difference between a support call and a user solving it in
 * four seconds. Sending people to Settings generically does not count — every
 * action here lands on the exact screen.
 *
 * Android can deep-link precisely. iOS cannot; see AirfobRemediation.swift for
 * what that costs.
 */
object AirfobRemediation {

    /** Action ids are shared with iOS and with the JS layer. */
    const val OPEN_BLUETOOTH = "openBluetoothSettings"
    const val OPEN_APP_SETTINGS = "openAppSettings"
    const val OPEN_LOCATION = "openLocationSettings"
    const val OPEN_NOTIFICATIONS = "openNotificationSettings"
    const val OPEN_BATTERY = "openBatterySettings"
    const val REQUEST_BATTERY_EXEMPTION = "requestBatteryExemption"

    /**
     * @return true when a settings screen was opened.
     * @throws AirfobError when the id is unknown or nothing on the device can
     *   handle it — callers surface that rather than leaving a dead button.
     */
    fun run(context: Context, actionId: String): Boolean {
        val intent = intentFor(context, actionId)
            ?: throw AirfobError("E_SDK", "Unknown remediation action \"$actionId\"")

        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

        val started = tryStart(context, intent)
        if (started) {
            AirfobLog.write("info", "perm", "REMEDIATE", "Opened settings for $actionId")
            return true
        }

        // OEM builds routinely drop individual settings screens. App details is
        // present on every device and gets the user to the right general area.
        val fallback = appDetailsIntent(context).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        if (actionId != OPEN_APP_SETTINGS && tryStart(context, fallback)) {
            AirfobLog.write(
                "warn", "perm", "REMEDIATE_FALLBACK",
                "No handler for $actionId — opened app details instead"
            )
            return true
        }

        AirfobLog.write("error", "perm", "REMEDIATE_FAIL", "Nothing could handle $actionId")
        throw AirfobError("E_SDK", "No settings screen on this device can handle \"$actionId\"")
    }

    private fun intentFor(context: Context, actionId: String): Intent? = when (actionId) {
        OPEN_BLUETOOTH -> Intent(Settings.ACTION_BLUETOOTH_SETTINGS)

        OPEN_APP_SETTINGS -> appDetailsIntent(context)

        OPEN_LOCATION -> Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)

        OPEN_NOTIFICATIONS ->
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS)
                    .putExtra(Settings.EXTRA_APP_PACKAGE, context.packageName)
            } else {
                appDetailsIntent(context)
            }

        // The list, not the direct prompt. See below.
        OPEN_BATTERY -> Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS)

        /**
         * The one-tap exemption prompt. Requires REQUEST_IGNORE_BATTERY_OPTIMIZATIONS
         * in the manifest, which this package deliberately does NOT declare:
         * Google Play restricts that permission to apps whose core function needs
         * it, and shipping it without an approved declaration risks the listing.
         *
         * Access control plausibly qualifies. If your listing is approved, add the
         * permission to your app manifest and this action starts working; until
         * then OPEN_BATTERY is the safe route.
         */
        REQUEST_BATTERY_EXEMPTION ->
            Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS)
                .setData(Uri.parse("package:" + context.packageName))

        else -> null
    }

    private fun appDetailsIntent(context: Context) =
        Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
            .setData(Uri.fromParts("package", context.packageName, null))

    private fun tryStart(context: Context, intent: Intent): Boolean =
        runCatching { context.startActivity(intent); true }.getOrDefault(false)
}
